### PR TITLE
AGENT-874: use AddNodesConfig asset

### PR DESCRIPTION
**Pull request by andfasano**
This patch is based on PR #8009.

The `nodes-config.yaml` is managed by the `AddNodesConfig` asset, and used in the asset graph. 
First injection of `ClusterInfo` and `AddNodesConfig` assets into the `Ignition` asset.

### CODE REVIEW

File: pkg/asset/agent/manifests/nmstateconfig.go [hunk 1]
```diff
@@ -80,8 +80,7 @@ func (n *NMStateConfig) Generate(dependencies asset.Parents) error {
 	installConfig := &agent.OptionalInstallConfig{}
 	dependencies.Get(agentHosts, installConfig, agentWorkflow, clusterInfo)
 
-	// Temporary skip in case of add nodes workflow. Main input for this
-	// asset will be retrieved from the nodes-config.yaml asset
+	// Not required for the add nodes workflow.
```
Comment (rwsu) [inline pkg/asset/agent/manifests/nmstateconfig.go:83]: I'm having trouble understanding why we wouldn't want the nmstateconfigs to be generated if they were provided in nodes-config.yaml.

File: pkg/asset/agent/image/ignition.go [hunk 2]
```diff
@@ -292,16 +318,16 @@ func addBootstrapScripts(config *igntypes.Config, releaseImage string) (err erro
 
 func getTemplateData(name, pullSecret, releaseImageList, releaseImage,
 	releaseImageMirror string, haveMirrorConfig bool, publicContainerRegistries string,
-	agentClusterInstall *hiveext.AgentClusterInstall,
+	numMasters int, numWorkers int,
```
Comment (bfournie) [inline pkg/asset/agent/image/ignition.go:321]: Nit...
```suggestion
	numMasters, numWorkers int,
```

File: pkg/asset/agent/image/ignition.go [hunk 3]
```diff
@@ -122,14 +130,42 @@ func (a *Ignition) Generate(dependencies asset.Parents) error {
...
+		numMasters = agentManifests.AgentClusterInstall.Spec.ProvisionRequirements.ControlPlaneAgents
+		numWorkers = agentManifests.AgentClusterInstall.Spec.ProvisionRequirements.WorkerAgents
+
+	case workflow.AgentWorkflowTypeAddNodes:
+		// In the add-nodes workflow, every node will act independently from the others.
+		a.RendezvousIP = "127.0.0.1"
+		// Reuse the existing cluster name.
+		clusterName = clusterInfo.ClusterName
+		// Fetch the required number of master and worker nodes.
+		numMasters = 0
```
Comment (bfournie) [inline pkg/asset/agent/image/ignition.go:162]: So for the release is replacing a master node not allowed? Maybe add a comment here regarding that.

File: pkg/asset/agent/joiner/addnodesconfig.go [hunk 4]
```diff
@@ -71,10 +87,44 @@ func (a *AddNodesConfig) Load(f asset.FileFetcher) (bool, error) {
...
+	return nil
+}
+
+func (a *AddNodesConfig) validateHosts() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if len(a.Config.Hosts) == 0 {
+		fieldPath := field.NewPath("hosts")
+		allErrs = append(allErrs, field.Required(fieldPath, "at least one host must be defined"))
+	}
```
Comment (bfournie) [inline pkg/asset/agent/joiner/addnodesconfig.go:127]: Will complete validation of the hosts info in the file be done here, including `staticNetworkConfigGenerator.ValidateStaticConfigParams` or will that be done somewhere else?

Comment (andfasano) [reply pkg/asset/agent/joiner/addnodesconfig.go:127]: Thanks @bfournie for raising up this point, on a second view the NMStateConfig asset if effectively required, I fixed it here https://github.com/openshift/installer/pull/8093/commits/b4c8aed7aa1930aa8c1fe28b5bbda7b3c5249195

**Review by rwsu — COMMENTED** [review]
/lgtm

File: pkg/asset/agent/manifests/nmstateconfig.go (hunk 1 above)
Comment (andfasano) [reply pkg/asset/agent/manifests/nmstateconfig.go:83]: You were right @rwsu, it's effectively required. I've fixed it in the next PR: https://github.com/openshift/installer/pull/8093

**Review by rwsu — COMMENTED** [review]
/lgtm

**Pull request merged**